### PR TITLE
implement mobile services on mar view

### PR DESCRIPTION
**Pull request by philbrookes**
Implement mobile services overview on MAR view.

Verification:
- provision mobile client
- provision non-mobile service
- provision keycloak and custom runtime connector
- in MAR click on Mobile Services tab
-- Confirm that keycloak and custom runtime connector are visible
-- Confirm that non-mobile service is not visible
- create binding (manually add `binding.aerogear.org/consumer` `binding.aerogear.org/provider` to binding)
-- Confirm that Mobile Services tab is updated
- deprovision keycloak and custom runtime connector
-- Confirm that link to catalog is displayed in Mobile Services tab

### CODE REVIEW

**Comment by aliok** [comment]
:eye: :eye:

**Comment by aliok** [comment]
screenshot from 2018-05-11 13-53-12

I was able to see links to Keycloak realms and UPS variants.
+1

**Review by aliok — APPROVED** [review]

File: app/scripts/controllers/mobileClients.js [hunk 1]
```diff
@@ -12,6 +12,7 @@ angular.module('openshiftConsole')
     $q,
     $routeParams,
     $scope,
+    $location,
```
Comment (sedroche) [inline app/scripts/controllers/mobileClients.js]: Is this used?

File: app/scripts/controllers/mobileClients.js [hunk 2]
```diff
@@ -57,6 +57,24 @@ angular.module('openshiftConsole')
                 }
                 ctrl.mobileClient = mobileClient;
               }));
+
+              watches.push(DataService.watch(APIService.getPreferredVersion("serviceinstances"), {namespace: ctrl.projectName}, function(serviceInstances){
```
Comment (sedroche) [inline app/scripts/controllers/mobileClients.js]: I think creating a higher level component that handles these watches and lists the services like we did for integrations might be good. It will keep the tab logic separate from the view controller.

File: app/scripts/controllers/mobileClients.js [hunk 3]
```diff
@@ -57,6 +57,24 @@ angular.module('openshiftConsole')
...
+                ctrl.serviceInstances = [];
+                _.each(serviceInstances.by('metadata.name'), function(serviceInstance){
```
Comment (sedroche) [inline app/scripts/controllers/mobileClients.js]: `_.filter`?

File: app/scripts/controllers/mobileClients.js [hunk 4]
```diff
@@ -57,6 +57,24 @@ angular.module('openshiftConsole')
...
+                  if ($filter('isMobileService')(ctrl.serviceClasses[serviceInstance.spec.clusterServiceClassRef.name])){
+                    ctrl.serviceInstances.push(serviceInstance);
+                  }
+                });
+              }));
+
+              watches.push(DataService.watch(APIService.getPreferredVersion("servicebindings"), {namespace: ctrl.projectName}, function(bindings){
+                ctrl.bindings = {};
+                _.each(bindings.by("metadata.name"), function(binding){
```
Comment (sedroche) [inline app/scripts/controllers/mobileClients.js]:
```.js
_(bindings.by("metadata.name"))
  .filter()
  .reduce()
  .value();
```
?

File: app/scripts/directives/mobileServiceInstanceRow.js [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+      'APIService',
+      'AuthorizationService',
+      'BindingService',
+      'Constants',
+      'DataService',
+      'ListRowUtils',
+      'ServiceInstancesService',
+      MobileServiceInstanceRow
+    ],
+    controllerAs: 'row',
```
Comment (sedroche) [inline app/scripts/directives/mobileServiceInstanceRow.js]: There is a default controllerAs of `$ctrl`

File: app/scripts/directives/mobileServiceInstanceRow.js [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+                              DataService,
+                              ListRowUtils,
+                              ServiceInstancesService) {
+    var row = this;
+    var serviceInstanceDisplayName = $filter('serviceInstanceDisplayName');
+    row.serviceBindingsVersion = APIService.getPreferredVersion('servicebindings');
+    row.state = row.state || {};
+    var annotationSpace = "org.aerogear.binding." + this.apiObject.metadata.name;
+    var watches = [];
+    _.extend(row, ListRowUtils.ui);
```
Comment (sedroche) [inline app/scripts/directives/mobileServiceInstanceRow.js]: What does this do?

File: app/scripts/directives/mobileServiceInstanceRow.js [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+                              AuthorizationService,
+                              BindingService,
+                              Constants,
+                              DataService,
+                              ListRowUtils,
+                              ServiceInstancesService) {
+    var row = this;
+    var serviceInstanceDisplayName = $filter('serviceInstanceDisplayName');
+    row.serviceBindingsVersion = APIService.getPreferredVersion('servicebindings');
+    row.state = row.state || {};
```
Comment (sedroche) [inline app/scripts/directives/mobileServiceInstanceRow.js]: Why is this needed?

File: app/scripts/directives/mobileServiceInstanceRow.js [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+    $scope.$on('$destroy', function(){
+        DataService.unwatchAll(watches);
+    });
+
+    watches.push(DataService.watchObject(Constants.MOBILE_CLIENT_VERSION, $routeParams.mobileclient, {namespace: $routeParams.project}, function(mobileClient, action) {
+        row.mobileClient = mobileClient;
+        row.annotations = [];
+        row.parameterData = {CLIENT_ID: row.mobileClient.metadata.name};
+        _.each(row.mobileClient.metadata.annotations, function(annotation, name){
```
Comment (sedroche) [inline app/scripts/directives/mobileServiceInstanceRow.js]:
```.js
_(row.mobileClient.metadata.annotations)
  .filter()
  .map(JSON.parse)
  .value()
```
?

File: app/scripts/directives/mobileServiceInstanceRow.js [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+
+    var getServicePlan = function() {
+      var servicePlanName = ServiceInstancesService.getServicePlanNameForInstance(row.apiObject);
+      return _.get(row, ['state', 'servicePlans', servicePlanName]);
+    };
+
+    row.$doCheck = function() {
+      row.servicePlan = getServicePlan();
+      row.displayName = serviceInstanceDisplayName(row.apiObject, row.serviceClass);
+      row.isBindable = BindingService.isServiceBindable(row.apiObject, row.serviceClass, row.servicePlan);
```
Comment (sedroche) [inline app/scripts/directives/mobileServiceInstanceRow.js]: A mobile service is also identified as bindable to a client if the service class has the tag `mobile-client-enabled`. It might be an easier check.

File: app/scripts/directives/mobileServiceInstanceRow.js [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+            }
+        });
+    }));
+
+    var getServicePlan = function() {
+      var servicePlanName = ServiceInstancesService.getServicePlanNameForInstance(row.apiObject);
+      return _.get(row, ['state', 'servicePlans', servicePlanName]);
+    };
+
+    row.$doCheck = function() {
```
Comment (sedroche) [inline app/scripts/directives/mobileServiceInstanceRow.js]: why do we need a $doCheck can it be moved to onchanges?

File: app/scripts/directives/mobileServiceInstanceRow.js [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+      row.servicePlan = getServicePlan();
+      row.displayName = serviceInstanceDisplayName(row.apiObject, row.serviceClass);
+      row.isBindable = BindingService.isServiceBindable(row.apiObject, row.serviceClass, row.servicePlan);
+    };
+    
+    row.actionsDropdownVisible = function() {
```
Comment (sedroche) [inline app/scripts/directives/mobileServiceInstanceRow.js]: I would prefer these in the html. `canI` can be a filter is there one for `deletionTimestamp`?

File: app/styles/_mobile-clients.less [hunk 12]
```diff
@@ -6,4 +6,22 @@
       max-height: 350px;
     }
   }
+}
+
+.create-binding {
```
Comment (sedroche) [inline app/styles/_mobile-clients.less]: CSS should be alphabetical

File: app/styles/_mobile-clients.less [hunk 13]
```diff
@@ -6,4 +6,22 @@
...
+    font-size: 4em;
+    margin: auto;
+    clear: both;
+  }
+  .note {
+    font-size: 1.3em;
+  }
+}
+
+.note {
```
Comment (sedroche) [inline app/styles/_mobile-clients.less]: Can you namespace this? It's a bit general here. or should it be under `.create-binding` as well?

File: app/views/browse/mobile-clients.html [hunk 14]
```diff
@@ -60,6 +60,23 @@ <h3>Mobile Client Config</h3>
...
+                    ng-repeat="serviceInstance in ctrl.serviceInstances track by (serviceInstance | uid)"
+                    api-object="serviceInstance" 
+                    binding="ctrl.bindings[serviceInstance.metadata.name]"
+                    service-class="ctrl.serviceClasses[serviceInstance.spec.clusterServiceClassRef.name]">
+                  </mobile-service-instance-row>
+                </div>
+              </div>
+              <div class="note" ng-if="!(ctrl.serviceInstances | size)">
+                <p>Provision a mobile service to integrate it with your mobile client.</p>
+                <a class="btn btn-primary btn-lg" ng-href="project/{{ctrl.projectName}}/catalog?category=mobile&subcategory=services">Browse Catalog</a>
```
Comment (sedroche) [inline app/views/browse/mobile-clients.html]: they prefer to use navigate e.g.  https://github.com/aerogear/origin-web-console/commit/c151097753861c751b81103f2893bd2ceef29644#diff-4cc5d017ab8a77ecfad04db4f53e2066R66

File: app/scripts/directives/mobileServiceInstanceRow.js [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+    _.extend(row, ListRowUtils.ui);
+
+    $scope.$on('$destroy', function(){
+        DataService.unwatchAll(watches);
+    });
+
+    watches.push(DataService.watchObject(Constants.MOBILE_CLIENT_VERSION, $routeParams.mobileclient, {namespace: $routeParams.project}, function(mobileClient, action) {
+        row.mobileClient = mobileClient;
+        row.annotations = [];
+        row.parameterData = {CLIENT_ID: row.mobileClient.metadata.name};
```
Comment (sedroche) [inline app/scripts/directives/mobileServiceInstanceRow.js]: How we are prepopulating the params is here:

https://github.com/aerogear/origin-web-console/commit/b4048e47814f8c7249dd15bdd91b42185562411e#diff-0572175f954d118f1869e17803496186R72

https://github.com/aerogear/origin-web-console/commit/b4048e47814f8c7249dd15bdd91b42185562411e#diff-0572175f954d118f1869e17803496186R129

File: app/views/_mobile-service-instance-row.html [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+                <a ng-if="annotation.type === 'href'" href="{{annotation.value}}">{{ annotation.value }}</a>
+                <span ng-if="annotation.type === 'string'">{{ annotation.value }}</span>
+            </li>
+        </ul>
+        <span class="note" ng-if="row.binding && (row.annotations | size) == 0">No configuration data to show for this service.</span>
+        <div ng-if="!row.binding" class="row create-binding">
+            <div class="col-md-12">
+                <span class="fa fa-info-circle icon"></span>
+                <div class="note">Bind to this service to use it in your mobile application.</div>
+                <button class="btn btn-primary btn-lg" ng-click="row.showOverlayPanel('bindService', {target: row.apiObject})">Create Binding</button>
```
Comment (sedroche) [inline app/views/_mobile-service-instance-row.html]: mobile bindings is one for every service bar ups which allows 2. can we set a limit on those?

File: app/views/_mobile-service-instance-row.html [hunk 17]
```diff
@@ -0,0 +1,77 @@
+<div class="list-pf-item provisioned-service" ng-class="{ active: row.expanded }">
```
Comment (sedroche) [inline app/views/_mobile-service-instance-row.html:1]: The view should be accessible.

**Review by sedroche — CHANGES_REQUESTED** [review]

**Comment by pb82** [comment]
@aliok Nice, does the link work? Does it expand the variant when you click it?

File: app/scripts/directives/mobileServiceInstanceRow.js (hunk 10 above)
Comment (philbrookes) [reply app/scripts/directives/mobileServiceInstanceRow.js]: A lot of these came from the serviceInstanceRow that I copied.

File: app/scripts/directives/mobileServiceInstanceRow.js (hunk 11 above)
Comment (philbrookes) [reply app/scripts/directives/mobileServiceInstanceRow.js]: This was copied from the serviceInstanceRow.

File: app/views/_mobile-service-instance-row.html (hunk 17 above)
Comment (philbrookes) [reply app/views/_mobile-service-instance-row.html:1]: I'm not sure what I need to change here, could you elaborate?

File: app/scripts/directives/mobileServiceInstanceRow.js (hunk 5 above)
Comment (philbrookes) [reply app/scripts/directives/mobileServiceInstanceRow.js]: this is copied from serviceInstanceRow

File: app/scripts/directives/mobileServiceInstanceRow.js (hunk 6 above)
Comment (philbrookes) [reply app/scripts/directives/mobileServiceInstanceRow.js]: this is copied from serviceInstanceRow, removing it seems to stop the expanding and collapsing from working, and the kebab menu.

File: app/scripts/directives/mobileServiceInstanceRow.js (hunk 7 above)
Comment (philbrookes) [reply app/scripts/directives/mobileServiceInstanceRow.js]: state is expected by the ListRowUtils.ui, and is not always set, this initiates it to an empty object if it is not passed in.

**Comment by aliok** [comment]
@pb82 yes it does. Link works fine.

File: app/views/_mobile-service-instance-row.html (hunk 16 above)
Comment (philbrookes) [reply app/views/_mobile-service-instance-row.html]: I'm not sure the best way to implement this, do you think it should be a special annotation on the mobile-client? 

How should the value of the limit be communicated to the UI from the APB?

**Comment by pb82** [comment]
@aliok I've tried this PR with a UPS variant but it doesn't show for me even though the annotation is there:

bildschirmfoto 2018-05-14 um 13 11 26

I can't see the variant in the `Mobile Services` tab. Anything i'm missing?

**Comment by aliok** [comment]
@pb82 you need to do thing described as "binding.aerogear.org/consumer binding.aerogear.org/provider"

* add consumer annotation to UPS config with value of the name of the mobile client
* add provider annotation to mobile client config with the value of the name of the UPS service

I think it is best if I show you over Bjeans

File: app/scripts/directives/mobileServiceInstanceRow.js (hunk 6 above)
Comment (sedroche) [reply app/scripts/directives/mobileServiceInstanceRow.js]: :+1:

File: app/scripts/controllers/mobileClients.js [hunk 18]
```diff
@@ -73,6 +76,24 @@ angular.module('openshiftConsole')
...
+                ctrl.serviceInstances = [];
+                ctrl.serviceInstances = _.filter(serviceInstances.by('metadata.name'), function(serviceInstance){
+                  return $filter('isMobileService')(ctrl.serviceClasses[serviceInstance.spec.clusterServiceClassRef.name]);
+                });
+              }));
+
+              watches.push(DataService.watch(APIService.getPreferredVersion("servicebindings"), {namespace: ctrl.projectName}, function(bindings){
+                ctrl.bindings = _(bindings.by("metadata.name"))
+                  .filter(function(binding){
+                    return _.get(binding.metadata.annotations, "binding.aerogear.org/consumer") === ctrl.mobileClient.metadata.name;
```
Comment (pb82) [inline app/scripts/controllers/mobileClients.js]: Are the APBs supposed to add those annotations to the service binding?

File: app/styles/_mobile-clients.less [hunk 19]
```diff
@@ -1,6 +1,28 @@
-.mobile-clients-view .copy-to-clipboard {
-  margin-bottom: 25px;
-  pre {
-    max-height: 350px;
+
+.create-binding {
+  text-align: center;
+  .icon {
+    width: 64px;
+    height: 64px;
```
Comment (dimitraz) [inline app/styles/_mobile-clients.less]: Is it possible to remove this height and add equal top/bottom padding to `.note`?

File: app/views/_mobile-service-instance-row.html [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+            <li ng-repeat="annotation in row.annotations">
+                <b>{{annotation.label}}</b>: 
+                <a ng-if="annotation.type === 'href'" href="{{annotation.value}}">{{ annotation.value }}</a>
+                <span ng-if="annotation.type === 'string'">{{ annotation.value }}</span>
+            </li>
+        </ul>
+        <span class="note" ng-if="row.binding && (row.annotations | size) == 0">No configuration data to show for this service.</span>
+        <div ng-if="!row.binding" class="row create-binding">
+            <div class="col-md-12">
+                <span class="fa fa-info-circle icon"></span>
```
Comment (dimitraz) [inline app/views/_mobile-service-instance-row.html]: Can this icon be changed to the patternfly version (`pficon-info`) with a colour of `#888` to match the mockups?

File: app/scripts/controllers/mobileClients.js [hunk 21]
```diff
@@ -49,6 +49,10 @@ angular.module('openshiftConsole')
         });
       };
 
+      ctrl.goToCatalog = function() {
```
Comment (sedroche) [inline app/scripts/controllers/mobileClients.js]: Move to `mobileServiceInstanceList`?

File: app/scripts/controllers/mobileClients.js [hunk 22]
```diff
@@ -60,19 +64,8 @@ angular.module('openshiftConsole')
             DataService.get(APIService.getPreferredVersion('mobileclients'), $routeParams.mobileclient, context, { errorNotification: false })
           ]).then(_.spread(function(serviceClasses, mobileClient) {
               ctrl.loaded = true;
-
               ctrl.serviceClasses = serviceClasses.by('metadata.name');
               ctrl.mobileClient = mobileClient;
-
-              watches.push(DataService.watchObject(APIService.getPreferredVersion('mobileclients'), $routeParams.mobileclient, context, function(mobileClient, action) {
```
Comment (sedroche) [inline app/scripts/controllers/mobileClients.js]: Why was this removed?

File: app/styles/_overview.less [hunk 23]
```diff
@@ -353,6 +353,18 @@
       a {
         .word-break();
       }
+      .logo {
```
Comment (sedroche) [inline app/styles/_overview.less]: Why is this in `_overview.less`? also needs to be alphabetical

File: app/scripts/directives/mobileServiceInstanceRow.js [hunk 24]
```diff
@@ -0,0 +1,124 @@
...
+        })
+        .map(JSON.parse)
+        .reduce(function(result, annotation){
+          result[_.get(annotation, 'label').trim()] = annotation;
+          return result;
+        }, {});
+        _(unsortedAnnotations).keys().sort().each(function (key) {
+          row.annotations[key] = unsortedAnnotations[key];
+        });       
+        watches.push(DataService.watch(APIService.getPreferredVersion("servicebindings"), {namespace: $routeParams.project}, function(bindings){
```
Comment (sedroche) [inline app/scripts/directives/mobileServiceInstanceRow.js]: Does this watch get set on every mobileClient change?

File: app/scripts/directives/mobileServiceInstanceRow.js [hunk 25]
```diff
@@ -0,0 +1,124 @@
...
+        var unsortedAnnotations = _.filter(row.mobileClient.metadata.annotations, function(annotation, name){
+          return name.split("/")[0] === annotationSpace;
+        })
+        .map(JSON.parse)
+        .reduce(function(result, annotation){
+          result[_.get(annotation, 'label').trim()] = annotation;
+          return result;
+        }, {});
+        _(unsortedAnnotations).keys().sort().each(function (key) {
+          row.annotations[key] = unsortedAnnotations[key];
```
Comment (sedroche) [inline app/scripts/directives/mobileServiceInstanceRow.js]: Does this do the same thing?
```.js
        var annotationCompare = function(a, b) {
          if (a.label < b.label) {
            return -1;
          }
          if (a.label > b.label) {
            return 1;
          }
          // a must be equal to b
          return 0;
        }
        row.annotations = _.filter(row.mobileClient.metadata.annotations, function(annotation, name){
          return name.split("/")[0] === annotationSpace;
        })
        .map(JSON.parse)
        .sort(annotationCompare);
```

File: app/scripts/directives/mobileServiceInstanceRow.js [hunk 26]
```diff
@@ -0,0 +1,124 @@
...
+        });       
+        watches.push(DataService.watch(APIService.getPreferredVersion("servicebindings"), {namespace: $routeParams.project}, function(bindings){
+          row.bindings = _.filter(bindings.by("metadata.name"), function(binding){
+            return _.get(binding.metadata.annotations, "binding.aerogear.org/consumer") === row.mobileClient.metadata.name &&
+                   _.get(binding.metadata.annotations, "binding.aerogear.org/provider") === row.apiObject.metadata.name;
+          });
+
+          function bindingInProgress(binding){
```
Comment (sedroche) [inline app/scripts/directives/mobileServiceInstanceRow.js]: There are some filters around binding state -> https://github.com/openshift/origin-web-common/blob/master/src/filters/resources.js#L155

maybe we can use those?

File: app/scripts/directives/mobileServiceInstanceRow.js [hunk 27]
```diff
@@ -0,0 +1,124 @@
...
+        .reduce(function(result, annotation){
+          result[_.get(annotation, 'label').trim()] = annotation;
+          return result;
+        }, {});
+        _(unsortedAnnotations).keys().sort().each(function (key) {
+          row.annotations[key] = unsortedAnnotations[key];
+        });       
+        watches.push(DataService.watch(APIService.getPreferredVersion("servicebindings"), {namespace: $routeParams.project}, function(bindings){
+          row.bindings = _.filter(bindings.by("metadata.name"), function(binding){
+            return _.get(binding.metadata.annotations, "binding.aerogear.org/consumer") === row.mobileClient.metadata.name &&
```
Comment (sedroche) [inline app/scripts/directives/mobileServiceInstanceRow.js]: To create a binding with these annotations you need to include this change in web common -> https://github.com/openshift/origin-web-common/pull/334 and pass `bindMeta` to the `bindService` -> https://github.com/aerogear/origin-web-console/commit/b4048e47814f8c7249dd15bdd91b42185562411e#diff-9b6377c25f28da45ed50c1d71ef2863aR248

**Review by sedroche — CHANGES_REQUESTED** [review]

**Comment by sedroche** [comment]
Can you make this more responsive?

responsive

File: app/scripts/directives/mobileServiceInstanceRow.js (hunk 27 above)
Comment (philbrookes) [reply app/scripts/directives/mobileServiceInstanceRow.js]: Done

File: app/scripts/directives/mobileServiceInstanceRow.js (hunk 25 above)
Comment (philbrookes) [reply app/scripts/directives/mobileServiceInstanceRow.js]: This is updated, I suppose github is getting confused.

File: app/scripts/directives/mobileServiceInstanceRow.js [hunk 28]
```diff
@@ -0,0 +1,124 @@
...
+
+        if(!bindingsWatch){
+          bindingsWatch = DataService.watch(APIService.getPreferredVersion("servicebindings"), {namespace: $routeParams.project}, function(bindings){
+            row.bindings = _.filter(bindings.by("metadata.name"), function(binding){
+              return _.get(binding.metadata.annotations, "binding.aerogear.org/consumer") === row.mobileClient.metadata.name &&
+                    _.get(binding.metadata.annotations, "binding.aerogear.org/provider") === row.apiObject.metadata.name;
+            });
+      
+            row.bindingInProgress = false;
+            _.each(row.bindings, function(binding){
```
Comment (sedroche) [inline app/scripts/directives/mobileServiceInstanceRow.js]: Can this filter be moved to the template?

File: app/scripts/directives/mobileServiceInstanceRow.js [hunk 29]
```diff
@@ -0,0 +1,124 @@
...
+    row.state = row.state || {};
+    var annotationSpace = "org.aerogear.binding." + this.apiObject.metadata.name;
+    var watches = [];
+    var bindingsWatch = false;
+    row.bindingsLimit = _.get(row.serviceClass, "spec.externalMetadata.bindingsLimit", 1);
+    _.extend(row, ListRowUtils.ui);
+
+    $scope.$on('$destroy', function(){
+        DataService.unwatchAll(watches);
+        DataService.unwatchAll([bindingsWatch]);
```
Comment (sedroche) [inline app/scripts/directives/mobileServiceInstanceRow.js]: `unwatch` will handle a single watch object

File: app/scripts/directives/overview/serviceInstanceRow.js [hunk 30]
```diff
@@ -4,6 +4,7 @@
   angular.module('openshiftConsole').component('serviceInstanceRow', {
     controller: [
       '$filter',
+      '$rootScope',
```
Comment (sedroche) [inline app/scripts/directives/overview/serviceInstanceRow.js]: I think this needs to be removed. Is it part of mobile-bindings?

File: app/styles/_mobile-clients.less [hunk 31]
```diff
@@ -1,6 +1,45 @@
-.mobile-clients-view .copy-to-clipboard {
-  margin-bottom: 25px;
-  pre {
-    max-height: 350px;
+
+.create-binding {
```
Comment (sedroche) [inline app/styles/_mobile-clients.less]: Can you namespace these under `.mobile-clients-view`

File: app/views/overview/_service-instance-row.html [hunk 32]
```diff
@@ -160,7 +160,7 @@ <h3>
         </div>
```
Comment (sedroche) [inline app/views/overview/_service-instance-row.html]: needs to be removed - part of mobile-bindings

**Review by sedroche — CHANGES_REQUESTED** [review]

File: app/scripts/directives/mobileServiceInstanceRow.js (hunk 28 above)
Comment (philbrookes) [reply app/scripts/directives/mobileServiceInstanceRow.js]: I don't think so, as it needs to toggle off in any of the bindings in the array of bindings are not ready. the ng-if would only compare against one binding, wouldn't it?

**Review by sedroche — APPROVED** [review]
